tools/libboard/board_n3000/board_n3000.c

```diff
@@ -313,25 +270,328 @@ fpga_result print_board_info(fpga_token token)
 	return res;
 }
 
+//enum eth group feature dfl_dev*
+fpga_result enum_eth_group_feature(fpga_token token,
+				   char eth_feature[ETH_GROUP_COUNT][SYSFS_MAX_SIZE],
+				   uint32_t size)
+{
+	fpga_result res          = FPGA_OK;
+	uint32_t i               = 0;
+	uint32_t index           = 0;
+	uint64_t value           = 0;
+	char sysfs_path[SYSFS_MAX_SIZE]     = { 0 };
+
+	for (i = 0; i < MAX_FPGA_FEATURE_COUNT; i++) {
+
+		value = 0;
+		if (snprintf(sysfs_path, SYSFS_MAX_SIZE,
+				"dfl_dev.%d/feature_id", i) < 0) {
+			OPAE_ERR("snprintf buffer overflow");
+			return FPGA_EXCEPTION;
+		}
+
+		res = read_sysfs_int64(token, sysfs_path, &value);
+		if (res != FPGA_OK) {
+			OPAE_MSG("Failed to read feature id");
+			continue;
+		}
+
+		if (value == ETH_GROUP_FEATURE_ID) {
+			if (snprintf(eth_feature[index], SYSFS_MAX_SIZE,
+				"dfl_dev.%d", i) < 0) {
+				OPAE_ERR("snprintf buffer overflow");
+				return FPGA_EXCEPTION;
+			}
+			index = index + 1;
+		}
+
+		if (index >= size) {
+			break;
+		}
+	}
+
+	if (index  < size) {
+		OPAE_ERR("Not found eth group Feature");
+		return FPGA_NOT_FOUND;
+	}
+
+	return res;
+}
+
+//enum pkvl regmap path
+// /sys/kernel/debug/regmap/spi%d.0/registers
+fpga_result enum_pkvl_sysfs_path(fpga_token token, char *pkvl_path)
+{
+	char sysfs_path[SYSFS_MAX_SIZE] = { 0 };
+	fpga_result res                 = FPGA_OK;
+	uint32_t i                      = 0;
+	fpga_object fpga_object;
+
+	for (i = 0; i < MAX_FPGA_FEATURE_COUNT; i++) {
+
+		if (snprintf(sysfs_path, SYSFS_MAX_SIZE,
+			"dfl*/*spi*/spi_master/spi*/spi%d*/", i) < 0) {
+			OPAE_ERR("snprintf buffer overflow");
+			return FPGA_EXCEPTION;
+		}
+
+		res = fpgaTokenGetObject(token, sysfs_path,
+						&fpga_object, FPGA_OBJECT_GLOB);
+		if (res != FPGA_OK) {
+			OPAE_MSG("Failed to get token Object");
+			continue;
+		}
+
+		res = fpgaDestroyObject(&fpga_object);
+		if (res != FPGA_OK) {
+			OPAE_ERR("Failed to Destroy Object");
+		}
+
+		if (snprintf(pkvl_path, SYSFS_MAX_SIZE,
+					"/sys/kernel/debug/regmap/spi%d.0/registers", i) < 0) {
+			OPAE_ERR("snprintf buffer overflow");
+			return FPGA_EXCEPTION;
+		}
+
+		return res;
+
+	}
+	return FPGA_NOT_FOUND;
+}
+
+//enum pkvl regmap path
+fpga_result read_regmap(char *sysfs_path,
+			uint64_t index,
+			uint32_t *value)
+{
+	FILE *fp                          = NULL;
+	char *endptr                      = NULL;
+	uint64_t line_count               = 0;
+	char search_str[SYSFS_MAX_SIZE]   = { 0 };
+	char line[MAX_LINE_LENGTH]        = { 0 };
+
+	if (snprintf(search_str, SYSFS_MAX_SIZE,
+				"%lx", index) < 0) {
+		OPAE_ERR("snprintf buffer overflow");
+		return FPGA_EXCEPTION;
+	}
+
+	fp = fopen(sysfs_path, "rb");
```

Comment:
It's a text file.  Do you need "rb" or just "r"?

Reply:
fixed